File: array/Special-Index-After-Removing-an-Element.js
/*
Special-Index-After-Removing-an-Element, sum of all EVEN indexed elements is equals to sum of all the ODD indexed elements.
*/

function SpecialIndexAfterRemovingAnElement(A) {
  // Prefix sum
  const pSumE = [], pSumO = [];
  pSumE[0] = A[0]; pSumO[0] = 0;

  for (let index = 1; index < A.length; index++) {
    const element = A[index];
    if (index % 2 == 0) {
      pSumE[index] = element + pSumE[index - 1];

      pSumO[index] = pSumO[index - 1];
    } else {
      pSumE[index] = pSumE[index - 1];

      pSumO[index] = element + pSumO[index - 1];
    }
  }
  // console.log(pSumE);
  // console.log(pSumO);

  const R = [];
  const N = A.length;
  for (let index = 0; index < N; index++) {
    // const sumOfEvenIndeces = pSumE[index - 1] + (pSumO[N - 1] - pSumO[index]);
    const sumOfEvenIndeces = (pSumE[index - 1] || 0) + (pSumO[N - 1] - pSumO[index]);
    // console.log({ index, sumOfEvenIndeces });
    const sumOfOddIndeces = (pSumO[index - 1] || 0) + (pSumE[N - 1] - pSumE[index]);
    // console.log({ index, sumOfOddIndeces });
    if (sumOfEvenIndeces == sumOfOddIndeces) {
      R.push(index);
    }
  }
  return R;
}

function main() {
  const DATA = [
    {
      A: [2, 1, 6, 4],
      R: [1]
    },
    {
      A: [1, 2, 1],
      R: [1]
    },
    {
      A: [1, 2, 1, 1],
      R: [2, 3]
    },
    {
      A: [1, 1, 2, 1],
      R: [0, 1]
    }
  ];
  DATA.forEach(({ A, R }, index) => {
    const start = new Date();
    const Result = SpecialIndexAfterRemovingAnElement(A);
    const end = new Date();
    const ms = (end - start);
    if (R.toString() == Result.toString()) {
      console.log("Passed", index + 1, `in [${ms} ms]`);
    } else {
      console.log("Wrong Answer..!", index + 1, `in [${ms} ms]`);
      console.log("Expected", R);
      console.log("Result", Result);
    }
  });
}

main();

File: array/SpecialIndexAfterRemovingAnElement.java
import java.util.ArrayList;
import java.util.Arrays;

import helper.TestCaseArray;

public class SpecialIndexAfterRemovingAnElement {

  static TestCaseArray[] TestCases = {
      new TestCaseArray(new int[] { 1, 2 }, new int[] {}),
      new TestCaseArray(new int[] { 1, 2, 3 }, new int[] {}),
      new TestCaseArray(new int[] { 1, 2, 4, 6 }, new int[] {}),
      new TestCaseArray(new int[] { 2, 1, 4, 6 }, new int[] {}),
      new TestCaseArray(new int[] { 2, 1, 6, 4 }, new int[] { 1 }),
      new TestCaseArray(new int[] { 1, 2, 2 }, new int[] { 0 }),
      new TestCaseArray(new int[] { 2, 2, 1 }, new int[] { 2 }),
  };

  public static void main(String[] args) {
    System.out.println();

    int count = 1;
    for (TestCaseArray testCase : TestCases) {
      int[] expected = testCase.R_Array;
      int[] result = specialIndicesAfterRemovingAnElement(testCase.A);
      if (Arrays.equals(result, expected)) {
        System.out.println(count + " Test case Passed!");
      } else {
        System.out.println(count + " Test case failed!");
        System.out.println("Expected: " + Arrays.toString(expected) + ", Result: " + Arrays.toString(result) + "\n");
      }
      count++;
    }
  }

  /**
   * Calculate special indices after removing an element.
   * Sum of Even Indices == Sum of Odd Indices
   * 
   * @param A
   * @return
   */
  public static int[] specialIndicesAfterRemovingAnElement(int[] A) {
    int n = A.length;
    if (n == 0) {
      return new int[] {};
    }
    ArrayList<Integer> R = new ArrayList<>();
    int[] pSumEven = new int[n];
    int[] pSumOdd = new int[n];
    pSumEven[0] = A[0];
    pSumOdd[0] = 0;
    for (int i = 1; i < n; i++) {
      if (i % 2 == 0) {
        pSumEven[i] = A[i] + pSumEven[i - 1];
        pSumOdd[i] = pSumOdd[i - 1];
      } else {
        pSumOdd[i] = A[i] + pSumOdd[i - 1];
        pSumEven[i] = pSumEven[i - 1];
      }
    }
    for (int i = 0; i < n; i++) {
      int evenSum = 0;
      int oddSum = 0;
      if (i == 0) {
        evenSum = (pSumOdd[n - 1] - pSumOdd[i]);
        oddSum = (pSumEven[n - 1] - pSumEven[i]);
      } else {
        // Sum of even indices excluding i
        evenSum = pSumEven[i - 1] + (pSumOdd[n - 1] - pSumOdd[i]);
        oddSum = pSumOdd[i - 1] + (pSumEven[n - 1] - pSumEven[i]);
      }
      if (evenSum == oddSum) {
        R.add(i);
      }
    }

    return R.stream().mapToInt(i -> i).toArray();
  }

}
